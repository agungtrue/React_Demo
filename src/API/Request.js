const data = {
    "data": [
        {
            "user_id": "f97cb2a2",
            "user_name": "John Doe",
            "email": "john@doe-club.org",
            "score": 12,
            "registered": "2021-06-23T06:32:23+00:00"
        },
        {
            "user_id": "280ae5cb",
            "user_name": "Jane Monroe",
            "score": 8,
            "email": "jane@monroe.com",
            "registered": "2021-07-01T13:26:07+00:00"
        },
        {
            "user_id": "280ae5cc",
            "user_name": "Cari Stacey",
            "score": 1,
            "email": "cari.stacey@reactdev.io",
            "registered": "2021-05-12T23:59:22+00:00"
        }
    ]
}

const error = {
    "message": "Something went wrong :(",
    "data": []
}

const expectError = false
export const getData = new Promise((resolve, reject) => {
    setTimeout(() => {
        if(expectError) return reject(error)
        resolve(data);
    }, 1500);
})

export const updateData = (newchanges, allData) => {
    return new Promise((resolve, reject) => {
        if(expectError) return reject(error)

        // i assume dataList always true/not null
        if(allData && allData.length) {
            const dataList = allData;
            const changes = newchanges
            const checker = [...dataList].filter(data => data.user_id === changes.user_id);
    
            if(!checker) alert('data not found')
    
            // acting like user_id
            const findIndex = dataList.findIndex(data => data.user_id === changes.user_id);
    
            // save data
            dataList[findIndex] = { ...changes }

            // response
            resolve(dataList);
        }
    })
}
