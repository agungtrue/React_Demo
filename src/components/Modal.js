import React, { useState, useEffect } from "react";
import Button from 'react-bootstrap/Button'
import Modal from 'react-bootstrap/Modal'
import Form from 'react-bootstrap/Form'


const ModalName = (props) => {

    const { showed, handleModal, dataList, handleNewData } = props

    const [show, setShow] = useState(false);
    const handleClose = () => {
        setShow(false)
        handleModal(false)
    };

    const [newData, setNewData] = useState({
        user_name: '',
        email: '',
        score: ''
    })

    const [errorForm, setErrorForm] = useState({
        user_name: { text: '', show: false },
        email: { text: '', show: false },
        score: { text: '', show: false }
    })


    const createNewData = (data) => {
        let newUserInput = data
        newUserInput.user_id = generateId(8)

        const d = new Date()
        const year = d.getFullYear()
        const date = d.getDate() 
        const month = d.getMonth() 

         let dateTemp = `${year}-${month+1}-${date}T23:59:22+00:00`
        newUserInput.registered = dateTemp

        // merge
        let currentData = [...dataList]
        const addingData = [...currentData, {...newUserInput}]

        // pass to parent
        handleNewData(addingData)

         // reset form
         let reset = {
            user_name: { text: '', show: false },
            email: { text: '', show: false },
            score: { text: '', show: false }
        }
        setErrorForm(reset)

        let defaultInput = {
            user_name: '',
            email: '',
            score: ''
        }
        setNewData(defaultInput)
        handleClose()
    }

    const validateForm = () => {
        let valid = true
        let err = {}
        
        if(!newData.user_name || !newData.email || !newData.score) {
            valid = false

            if (!newData.user_name) err = {...err, user_name: { text: 'Username is mandatory', show: true}}
            if (!newData.email) err = {...err, email: { text: 'Email is mandatory', show: true}}
            if (!newData.score) err = {...err, score: { text: 'Score is mandatory', show: true}}

            console.log('err', err)
            setErrorForm(err)
        }

        if(newData.email) {
            let emailCheck = validateEmail(newData.email)
            if(!emailCheck) {
                err = {...err, email: { text: 'Email is not valid', show: true}}
                valid = false
                setErrorForm(err)
            }
        }

        if(newData.score) {
            let scoreCheck = validateScore(newData.score)
            if(!scoreCheck) {
                err = {...err, score: { text: 'Score must be between 0 and 100', show: true}}
                valid = false
                setErrorForm(err)
            }
        }

        // reset err
        if(valid) {
            let reset = {
              user_name: { text: '', show: false },
              email: { text: '', show: false },
              score: { text: '', show: false }
            }
            setErrorForm(reset)
        }

        return valid
    }

    const validateEmail = (email) => {
        let valid = false
        if(email.includes('@')) valid = true

        return valid
    }

    const validateScore = (score) => {
        let valid = false
        const limit = 100
        if(score.length <= 3 && Number(score) <= limit) valid = true

        return valid
    }

    const generateId = (length) => {
        let result           = '';
        let characters       = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        let charactersLength = characters.length;
        for ( var i = 0; i < length; i++ ) {
          result += characters.charAt(Math.floor(Math.random() * 
     charactersLength));
       }
       return result;
    }


    useEffect(() => {
        setShow(showed)
    }, [showed])

    return (
        <>
          <Modal show={show} onHide={handleClose}>
            <Modal.Header closeButton>
              <Modal.Title>Create Data</Modal.Title>
            </Modal.Header>
            <Modal.Body>
            <Form>
                <Form.Group className="mb-3" controlId="formBasicEmail">
                    <Form.Label>User Name</Form.Label>
                    <Form.Control 
                        type="text" 
                        placeholder="Enter Username" 
                        onChange={(e) => setNewData({...newData, user_name: e.target.value})} 
                    />
                    <div style={{color: 'red'}}> {errorForm.user_name && errorForm.user_name.show ? errorForm.user_name.text : ''} </div>
                </Form.Group>
                <Form.Group className="mb-3" controlId="formBasicPassword">
                    <Form.Label>Email</Form.Label>
                    <Form.Control 
                        type="email" 
                        placeholder="Enter Email"
                        onChange={(e) => setNewData({...newData, email: e.target.value})} 
                    />
                    <div style={{color: 'red'}}> {errorForm.email && errorForm.email.show ? errorForm.email.text : ''} </div>
                </Form.Group>
                <Form.Group className="mb-3" controlId="formBasicPassword">
                    <Form.Label>Score</Form.Label>
                    <Form.Control 
                        type="number" 
                        placeholder="Enter Score"
                        onChange={(e) => setNewData({...newData, score: e.target.value})} 
                    />
                    <div style={{color: 'red'}}> {errorForm.score && errorForm.score.show ? errorForm.score.text : ''} </div>
                </Form.Group>
                </Form>
            </Modal.Body>
            <Modal.Footer>
              <Button variant="secondary" onClick={handleClose}>
                Close
              </Button>
              <Button variant="primary" onClick={() => validateForm() && createNewData(newData) }>
                Add
              </Button>
            </Modal.Footer>
          </Modal>
        </>
      );
}

export default ModalName