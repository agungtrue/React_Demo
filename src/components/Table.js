import React, { useState, useEffect } from "react";
import Table from 'react-bootstrap/Table'
import Button from 'react-bootstrap/Button'
import ModalName from './Modal'


const Tables = (props) => {
    const { dataSource } = props
    const [showBtn, setShowBtn] = useState({
        data: {}
    })
    const [dataList, setDataList] = useState([])
    const [dataSelected, setDataSelected] = useState({})
    const [actionType, setActionType] = useState('')
    const [showModal, setShowModal] = useState(false)

    const [sortedBy, setSortedBy] = useState({
        user_name: '',
        email: '',
        score: ''
    })

    const deleteData = (dataParams) => {
        const currentData = [...dataList]
        const afterFemoveData = currentData.filter(data => data.user_id !== dataParams.user_id)

        //set new data after remove
        setDataList(afterFemoveData)
    }

    useEffect(() => {

        // waiting request for api and set to dataList
        setDataList(dataSource)

    }, [dataSource]) //watching data changes


    const openModalCreate = (type, data = {}) => {

        //set action type
        setActionType(type)

        if(type === 'create') {
            setShowModal(true)
        }
        else if (type === 'update') {
            setShowModal(true)
            setDataSelected(data)
        }
        else console.log('invalid type')
    }

    const handleModal = (value) => {
        setShowModal(value)
        setDataSelected({})
    }

    const handleNewData = (data) => {
        // replace for new one
        setDataList(data)
    }

    const sorting = (key) => {

        if(sortedBy[key] === 'asc') {

            // can use reverse method
            // const data = [...dataList].reverse()
            const data = [...dataList]

            let result = []
            for (let i = data.length - 1; i >= 0; i--) {
                result.push(data[i])
            }
            setSortedBy({...sortedBy, [key]: 'desc'})
            setDataList(result)
        }
        else {
            const data = [...dataList].sort((a, b) => (a[key] < b[key]) ? -1 : ((a[key] > b[key]) ? 1 : 0));
            setSortedBy({...sortedBy, [key]: 'asc'})
            setDataList(data)
        }
    }

    return (
        <>
            <Button 
                variant="primary" 
                size="lg" 
                onClick={() => openModalCreate('create')}
                style={{marginBottom: '1rem'}}> Create
            </Button>
            <Table striped bordered hover responsive>
                <thead>
                    <tr>
                        <th>No</th>
                        <th style={{cursor:'pointer'}} onClick={() => sorting('user_name')}>User Name ({sortedBy.user_name})</th>
                        <th style={{cursor:'pointer'}} onClick={() => sorting('score')}>Score ({sortedBy.score})</th>
                        <th style={{cursor:'pointer'}} onClick={() => sorting('registered')}>Registered ({sortedBy.registered})</th>
                    </tr>
                </thead>
                <tbody>
                    {
                            dataList && dataList.length ? dataList.map((data, index) => 
                            (
                                <tr
                                    key={data.user_id}
                                    onMouseEnter={() => setShowBtn({...showBtn, data})}
                                    onMouseLeave={() => setShowBtn({...showBtn, data: {}})}
                                >
                                    <td>{index+1}</td>
                                    <td style={{cursor:'pointer'}} onClick={() => openModalCreate('update', data)}>{data.user_name}</td>
                                    <td>{data.score}</td>
                                    <td style={{display: 'flex', justifyContent: 'space-between', cursor: 'pointer'}}>
                                        {data.registered}
                                        { showBtn.data && showBtn.data.user_id === data.user_id && <span onClick={() => deleteData(data)}>Delete</span> }
                                    </td>
                                </tr>
                            )
                        ) : 'Please wait....'
                    }
                </tbody>
            </Table>



            {/* Modal */}
            <ModalName
                showed={showModal}
                handleModal={handleModal}
                handleNewData={handleNewData}
                dataList={dataList}
                dataSelected={dataSelected}
                actionType={actionType}
            />
      </>
    )
}

export default Tables