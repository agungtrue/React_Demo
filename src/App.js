import React, { useState, useEffect } from "react";
import { notification } from 'antd';
import { getData } from './API/Request'
import './App.css';
import Table from './components/Table'

function App() {
  const [dataList, setDataList] = useState([])

  useEffect(() => {
      getApi('getData')
  }, [])

  //api methods
  const getApi = async (type) => {

      try {
          if(type === 'getData') {
            const response = await getData
            console.log('response', response.data)

            if(response && response.data) {
                setDataList(response.data)
            }
          }
          else alert('invalid request type')

      } catch (err) {
          console.log('err', err.message)

          notification.error({
              message: err.message,
              duration: 5
          });
      }
  }


  return (
    <div className="App">
      <p>Test Screening React JS Hooks</p>
      <div className="container">
        <Table
            dataSource={dataList}
         />
      </div>
    </div>
  );
}

export default App;
