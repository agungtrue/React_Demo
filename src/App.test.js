import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from "enzyme";
import App from './App';
import Table from './components/Table';
import Modal from './components/Modal';
import { getData } from './API/Request'


test("renders without crashing", () => {
    const component = renderer.create(<App />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("renders table component without crashing", () => {
    const component = renderer.create(<Table />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("renders Modal component without crashing", () => {
    const component = renderer.create(<Modal />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("snapshot with all required props in Table Component", () => {

  // with this test, i saw new error: Each child in a list should have a unique "key" prop.
  // and also i fixed alreaday after saw that error console test
  let sample = [{
      user_id: "f97cb2a2",
      user_name: "John Doe",
      email: "john@doe-club.org",
      score: 12,
      registered: "2021-06-23T06:32:23+00:00"
  }]

  const component = renderer.create(<Table
      dataSource={sample}
  />);
  expect(component.toJSON()).toMatchSnapshot();

});