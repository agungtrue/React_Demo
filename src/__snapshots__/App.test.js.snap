// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot with all required props 1`] = `
Array [
  <button
    className="btn btn-primary btn-lg"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "1rem",
      }
    }
    type="button"
  >
     Create
  </button>,
  <div
    className="table-responsive"
  >
    <table
      className="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            No
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            User Name (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Score (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Registered (
            )
          </th>
        </tr>
      </thead>
      <tbody>
        Please wait....
      </tbody>
    </table>
  </div>,
]
`;

exports[`renders Modal component without crashing 1`] = `null`;

exports[`renders table component without crashing 1`] = `
Array [
  <button
    className="btn btn-primary btn-lg"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "1rem",
      }
    }
    type="button"
  >
     Create
  </button>,
  <div
    className="table-responsive"
  >
    <table
      className="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            No
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            User Name (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Score (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Registered (
            )
          </th>
        </tr>
      </thead>
      <tbody>
        Please wait....
      </tbody>
    </table>
  </div>,
]
`;

exports[`renders without crashing 1`] = `
<div
  className="App"
>
  <p>
    Test Screening React JS Hooks
  </p>
  <div
    className="container"
  >
    <button
      className="btn btn-primary btn-lg"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginBottom": "1rem",
        }
      }
      type="button"
    >
       Create
    </button>
    <div
      className="table-responsive"
    >
      <table
        className="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              No
            </th>
            <th
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              User Name (
              
              )
            </th>
            <th
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              Score (
              
              )
            </th>
            <th
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              Registered (
              )
            </th>
          </tr>
        </thead>
        <tbody>
          Please wait....
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`snapshot with all required props 1`] = `
Array [
  <button
    className="btn btn-primary btn-lg"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "1rem",
      }
    }
    type="button"
  >
     Create
  </button>,
  <div
    className="table-responsive"
  >
    <table
      className="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            No
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            User Name (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Score (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Registered (
            )
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <td>
            1
          </td>
          <td
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            John Doe
          </td>
          <td>
            12
          </td>
          <td
            style={
              Object {
                "cursor": "pointer",
                "display": "flex",
                "justifyContent": "space-between",
              }
            }
          >
            2021-06-23T06:32:23+00:00
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;

exports[`snapshot with all required props in Table Component 1`] = `
Array [
  <button
    className="btn btn-primary btn-lg"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "1rem",
      }
    }
    type="button"
  >
     Create
  </button>,
  <div
    className="table-responsive"
  >
    <table
      className="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            No
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            User Name (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Score (
            
            )
          </th>
          <th
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Registered (
            )
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <td>
            1
          </td>
          <td
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            John Doe
          </td>
          <td>
            12
          </td>
          <td
            style={
              Object {
                "cursor": "pointer",
                "display": "flex",
                "justifyContent": "space-between",
              }
            }
          >
            2021-06-23T06:32:23+00:00
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
